MudStack: Add parameter StretchItems to have certain children fill space

BieleckiLtd opened the pull request:
If the original PR #6010 was rejected due to conflicts, this update brings the previous code up to date with the current dev branch. If the rejection was for a different reason, please close this PR. Thank you.

## Description
Closes #4960

https://user-images.githubusercontent.com/57046415/208328064-3eafca62-f9b4-4f08-8a19-32e59e9d0f03.mp4

## How Has This Been Tested?
unit + visually

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

<!-- If you made any visual changes, provide screenshots of before/after, it its moving parts, please provide high quality gif, wemb or mp4 -->

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [x] I've added relevant tests.

BieleckiLtd committed:
Add StretchChildren API + docs + tests

danielchalmers reviewed (approved):
Great work - LGTM

henon commented:
I like the functionality, it is very useful, reminds me of `LastChildFill` in WPF's `DockPanel`. Before we merge it, we need to make sure we got the best naming for the parameter, the enum and its values in view of the other parameters that apply to the children: `AlignItems` and `Justify`. 

Also, we should consider LTR vs RTL. `First` and `Last` are culture specific, so in RTL mode `First` should stretch the right most child. And since we use `Start` and `End` for all other culture specific directions in the lib we should use those instead.

Naming suggestion: 

```cs
[Parameter]
public StretchItems StretchItems { get; set; }

public enum StretchItems {
  None,
  Start,
  End,
  StartAndEnd,
  Middle, 
  All,
}
```

Usage:

```html
<MudStack StretchItems="StretchItems.End">
  ...
</MudStack>
```

It might be good to add a `StartAndEnd` also.

ScarletKuro commented:
> I like the functionality, it is very useful

Yeah, I think this is very common functionality in the Xamarin, Flutter, MAUI, WPF.

BieleckiLtd committed:
Merge branch 'MudBlazor:dev' into feature/stack-manage-children

BieleckiLtd commented:
> [...] we need to make sure we got the best naming for the parameter, the enum and its values in view of the other parameters [...]

Taking this further, perhaps `StretchItems` with `Stretch` enum would be more appropriate?
```
<MudStack AlignItems="Align.Start" StretchItems="Stretch.End">
...
</MudStack>
```

BieleckiLtd commented:
> Oh sorry, you didn't make the changes yet. Make sure to use the `[CascadingParameter] public bool RightToLeft {get; set;}` to achive the correct order for Start and End

`MudStack` has `Reverse` parameter which does that. Should this be replaced with CascadingParameter?

BieleckiLtd commented:
This is a tricky one because RTL does not mean the same in `MudStack`. When the stack is vertical (column), `Reverse` means bottom-to-top, not right-to-left.

henon commented:
I see that mudstack already implements right to left logic. so I guess you won't have to? It will be automatically correct?

![image](https://github.com/MudBlazor/MudBlazor/assets/44090/39c26b39-b856-4e7d-935a-abd44fb1a80b)

henon commented:
> This is a tricky one because RTL does not mean the same in `MudStack`. When the stack is vertical (column), `Reverse` means bottom-to-top, not right-to-left.

Reverse is independent of RTL
![image](https://github.com/MudBlazor/MudBlazor/assets/44090/34e70327-7f18-4b41-9c02-cc258d310184)

Question is, what should be done with start and end when Reverse is true? I don't know

Commit:
Adjusted names, added StartAndEnd to Enum, corrected props descriptions in MudStack, added xml comments.

BieleckiLtd commented:
Whatever it is now feels natural in both LTR and RTL when reversed, and when not, so I'd leave it as is. I have changed parameter name to `StretchItems`, Enum name to `Stretch`, added `Stretch.StartAndEnd` + unit, corrected some other incorrect descriptions in `MudStack` and updated examples to use new generic MudChipSet.

henon commented on src/MudBlazor.Docs/Pages/Components/Stack/StackPage.razor:
```diff
@@ -96,14 +96,14 @@
             </SectionHeader>
-            <SectionContent Block FullWidth Code="StackStretchChildrenExample" ShowCode="false" DarkenBackground>
-                <StackStretchChildrenExample />
+            <SectionContent Block FullWidth Code="StackStretchItemsExample" ShowCode="false" DarkenBackground>
```
Please use `Code="@nameof(StackStretchItemsExample)"` and you may also add nameofs in the other sections

BieleckiLtd replied:
Sure, sorted.

henon commented:
Perfect, looks good. If you will add the nameofs real quick I'll merge this today.

Commit:
Updated StackPage docs page to use nameof's

BieleckiLtd commented:
> AlignItems="Align.Start

I have made a mistake here, it's `AlignItems.Start`. Should `Stretch` be `StretchItems`?

henon commented:
In theory there could be other uses of stretching, i.e. for images where you would need Stretch.Uniform, Stretch.X, Stretch.Y etc, so yes, it is the safest bet to rename the enum to `StretchItems`

Commit:
Renamed Stretch enum as StretchItems

henon commented:
Thanks @BieleckiLtd

onatsko commented:
Hello guys
Thanks for the great components, they help a lot in my work!

Is it possible to implement this functionality in v6?

henon commented:
v6 is on community support which means that those who need something backported from higher versions or bug fixes can do so. We'll review PRs and merge them and release a new version of v6.
